lib/pg_party/adapter_decorator.rb

```diff
@@ -157,27 +367,69 @@ def quote_collection(values)
     end
 
     def template_table_name(table_name)
-      "#{table_name}_template"
+      "#{parent_for_table_name(table_name, traverse: true) || table_name}_template"
     end
 
     def range_constraint_clause(start_range, end_range)
       "FROM (#{quote_collection(start_range)}) TO (#{quote_collection(end_range)})"
     end
 
+    def hash_constraint_clause(modulus, remainder)
+      "WITH (MODULUS #{modulus.to_i}, REMAINDER #{remainder.to_i})"
+    end
+
     def list_constraint_clause(values)
       "IN (#{quote_collection(values.try(:to_a) || values)})"
     end
 
+    def partition_by_clause(type, partition_key)
+      "PARTITION BY #{type.to_s.upcase} (#{quote_partition_key(partition_key)})"
+    end
+
     def uuid_function
       try(:supports_pgcrypto_uuid?) ? "gen_random_uuid()" : "uuid_generate_v4()"
     end
 
     def hashed_table_name(table_name, key)
-      "#{table_name}_#{Digest::MD5.hexdigest(key)[0..6]}"
+      return "#{table_name}_#{Digest::MD5.hexdigest(key)[0..6]}" if key
+
+      # use _default suffix for default partitions (without a constraint clause)
+      "#{table_name}_default"
+    end
+
+    def index_valid?(index_name)
+      select_values(
+        "SELECT relname FROM pg_class, pg_index WHERE pg_index.indisvalid = false AND "\
+          "pg_index.indexrelid = pg_class.oid AND relname = #{quote(index_name)}"
+      ).empty?
+    end
+
+    def generate_index_name(index_name, table_name)
+      "#{index_name}_#{Digest::MD5.hexdigest(table_name)[0..6]}"
+    end
+
+    def validate_supported_partition_type!(partition_type)
+      if (sym = partition_type.to_s.downcase.to_sym) && sym.in?(SUPPORTED_PARTITION_TYPES)
+        return if sym != :hash || postgres_11_plus?
+
+        raise NotImplementedError, 'Hash partitions are only available in Postgres 11 or higher'
+      end
+
+      raise ArgumentError, "Supported partition types are #{SUPPORTED_PARTITION_TYPES.join(', ')}"
+    end
+
+    def validate_default_partition_support!
+      return if postgres_11_plus?
+
+      raise NotImplementedError, 'Default partitions are only available in Postgres 11 or higher'
     end
 
     def supports_partitions?
       __getobj__.send(:postgresql_version) >= 100000
     end
+
+    def postgres_11_plus?
+      __getobj__.send(:postgresql_version) >= 110000
```

Comment:
Kind of a nitpick, but maybe we should add a `postgres_version` method

Reply:
Refactored in https://github.com/rkrage/pg_party/pull/54/commits/b1783ffe1b70217480c818fd67701a191ff90264 where I added a `postgres_major_version` method